packages/react-dom/src/client/ReactDOMHostConfig.js

```diff
@@ -1215,32 +1270,161 @@ export function errorHydratingContainer(parentContainer: Container): void {
   }
 }
 
-export function getInstanceFromNode(node: HTMLElement): null | Object {
-  return getClosestInstanceFromNode(node) || null;
-}
-
-export function preparePortalMount(portalInstance: Instance): void {
-  listenToAllSupportedEvents(portalInstance);
+export function acquireSingletonInstance(
+  type: string,
+  props: Props,
+  rootContainerInstance: Container,
+  hostContext: HostContext,
+): Instance {
+  if (__DEV__) {
+    // TODO: take namespace into account when validating.
+    const hostContextDev = ((hostContext: any): HostContextDev);
+    validateDOMNesting(type, null, hostContextDev.ancestorInfo);
+  }
+  const ownerDocument = getOwnerDocumentFromRootContainer(
+    rootContainerInstance,
+  );
+  // For the three persistent Host Components that exist in DOM it is necessary for there to
+  // always be a documentElement. With normal html parsing this will always be the case but
+  // with pathological manipulation the document can end up in a state where no documentElement
+  // exists. We create it here if missing so we can treat it as an invariant.
+  // It is important to note that thi dom mutation and others in this function happen
+  // in render rather than commit. This is tolerable because they only happen in degenerate cases
+  if (!ownerDocument.documentElement) {
+    ownerDocument.append(ownerDocument.createElement('html'));
+  }
+  switch (type) {
+    case 'html': {
+      // We ensure this exists just above
+      return ((ownerDocument.documentElement: any): HTMLHtmlElement);
+    }
+    case 'head': {
+      if (!ownerDocument.head) {
+        ownerDocument.insertBefore(
+          ownerDocument.createElement('head'),
+          ownerDocument.firstChild,
+        );
+      }
+      // We ensure this exists just above
+      return ((ownerDocument.head: any): HTMLHeadElement);
+    }
+    case 'body': {
+      if (!ownerDocument.body) {
+        ownerDocument.appendChild(ownerDocument.createElement('body'));
+      }
+      // We ensure this exists just above
+      return ((ownerDocument.body: any): HTMLBodyElement);
+    }
+    default: {
+      throw new Error(
+        'acquireSingletonInstance was called with an element type that is not supported. This is a bug in React.',
+      );
+    }
+  }
 }
 
-export function prepareScopeUpdate(
-  scopeInstance: ReactScopeInstance,
+export function commitSingletonPlacement(
+  type: string,
+  props: Props,
+  instance: Instance,
   internalInstanceHandle: Object,
 ): void {
-  if (enableScopeAPI) {
-    precacheFiberNode(internalInstanceHandle, scopeInstance);
+  switch (type) {
+    case 'html':
+    case 'head':
+    case 'body': {
```

Comment:
There's no reason for this check other than for an error message. We can make that a DEV only check then.